gather-scheduler.js

```diff
@@ -18,99 +18,114 @@
+    return new Promise((res, _) => {
+      // Enable emulation.
+      if (options.flags.emulateMobileDevice) {
+        return res(driver.beginEmulation());
+      }
+
+      // noop if no mobile emulation
+      res();
+    }).then(_ => {
+      return driver.cleanAndDisableBrowserCaches();
+    }).then(_ => {
       // Force SWs to update on load.
-      .then(_ => driver.forceUpdateServiceWorkers())
-
-      // Gather: setup phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.setup(options)))
-
-      // Enable tracing and network record collection.
-      .then(_ => driver.beginTrace())
-      .then(_ => driver.beginNetworkCollect())
-
-      // Gather: beforePageLoad phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.beforePageLoad(options)))
-
-      // Load the page (if the CLI / extension want it loaded).
-      .then(_ => {
-        if (loadPage) {
-          return driver.gotoURL(url, driver.WAIT_FOR_LOADED);
-        }
-
-        return Promise.resolve();
-      })
+      return driver.forceUpdateServiceWorkers();
+    });
+  }
 
-      // Gather: afterPageLoad phase
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterPageLoad(options)))
+  // Enable tracing and network record collection.
+  static beginPassiveCollection(driver) {
+    return driver.beginTrace().then(_ => {
+      return driver.beginNetworkCollect();
+    });
+  }
 
-      // Disable network collection; grab records.
-      .then(_ => driver.endNetworkCollect())
-      .then(networkRecords => {
-        tracingData.networkRecords = networkRecords;
-      })
+  static endPassiveCollection(driver, gatherers, options, tracingData) {
+    return driver.endNetworkCollect().then(networkRecords => {
+      tracingData.networkRecords = networkRecords;
+    }).then(_ => {
+      return driver.endTrace();
+    }).then(traceContents => {
+      tracingData.traceContents = traceContents;
+    });
+  }
 
-      // Disable tracing; grab records.
-      .then(_ => driver.endTrace())
-      .then(traceContents => {
-        tracingData.traceContents = traceContents;
-      })
+  static _phase(phaseName) {
+    return function(driver, gatherers, options, tracingData) {
+      return GatherScheduler._runPhase(gatherers, gatherer => {
+        return gatherer[phaseName](options, tracingData);
+      });
+    };
+  }
 
-      // Gather: afterTraceCollected phase.
-      .then(_ => this._runPhase(gatherers,
-          gatherer => gatherer.afterTraceCollected(options, tracingData)))
+  static _runPhase(gatherers, gatherFun) {
+    return gatherers.reduce((chain, gatherer) => {
+      return chain.then(_ => gatherFun(gatherer));
+    }, Promise.resolve());
+  }
 
-      // Disconnect the driver.
-      .then(_ => driver.disconnect())
+  static run(gatherers, options) {
+    const driver = options.driver;
+    const tracingData = {};
+    const self = GatherScheduler;
```

Comment:
ok let's do this in another PR seems like a good abstraction / improvement.. or at least one that we can play around with in another PR sandbox